Fix missing analyzer warning for generics in containing type

sbomer opened the pull request:
Fixes the IL2091 warnings missed by the analyzer in https://github.com/dotnet/runtime/issues/109270.

Fixes https://github.com/dotnet/runtime/issues/109270

sbomer committed:
Fix missing analyzer warning for generics in containing type

jtschuster commented on src/tools/illink/src/ILLink.RoslynAnalyzer/TrimAnalysis/GenericArgumentDataFlow.cs:
```diff
@@ -21,12 +21,20 @@ public static void ProcessGenericArgumentDataFlow (Location location, IMethodSym
 
-			ProcessGenericArgumentDataFlow (location, method.ContainingType, reportDiagnostic);
+			var type = method.ContainingType;
+			while (type is { IsGenericType: true }) {
```
Same as below
```suggestion
			while (type.IsGenericType) {
```

jtschuster replied:
Also, would it make sense to recurse up the `ContainingType`s in the `INamedTypeSymbol` overload of `ProcessGenericArgumentDataflow` rather than in both the `IMethodSymbol` and `IFieldSymbol` overloads, or does doing that cause issues elsewhere?

jtschuster commented on src/tools/illink/test/Mono.Linker.Tests.Cases/DataFlow/GenericParameterDataFlow.cs:
```diff
@@ -849,6 +850,31 @@ static void TestNoWarningsInRUCType<T, U> ()
 			rucType.VirtualMethodRequiresPublicMethods<T> ();
 		}
 
+		class TestGenericParameterFlowsToNestedType
+		{
+			class Generic<T> {
+				[ExpectedWarning ("IL2091")]
+				public T CallNestedMethod () => GenericRequires<T>.Nested.Method ();
+
+				[ExpectedWarning ("IL2091")]
+				public T AccessNestedField () => GenericRequires<T>.Nested.Field;
+			}
+
+			class GenericRequires<[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.All)] T> {
+				public static class Nested {
+					public static T? Method () => default;
+
+					public static T? Field = default;
+				}
+			}
+
+			public static void Test ()
+			{
+				new Generic<int> ().CallNestedMethod ();
+				new Generic<string> ().AccessNestedField ();
+			}
+		}
+
```
Are properties and events relevant here?

sbomer committed:
PR feedback

- Add tests for property/event
- Fix behavior for type arguments and base types

jtschuster reviewed (approved):
Thank you!
